src/flamegraph/color/mod.rs

```diff
@@ -0,0 +1,220 @@
+use rand::rngs::ThreadRng;
+use rand::Rng;
+use std::str::FromStr;
+
+mod palette_map;
+mod palettes;
+
+pub(super) use palette_map::PaletteMap;
+
+pub(super) const VDGREY: &str = "rgb(160,160,160)";
+pub(super) const DGREY: &str = "rgb(200,200,200)";
+
+const YELLOW_GRADIENT: (&str, &str) = ("#eeeeee", "#eeeeb0");
+const BLUE_GRADIENT: (&str, &str) = ("#eeeeee", "#e0e0ff");
+const GRAY_GRADIENT: (&str, &str) = ("#f8f8f8", "#e8e8e8");
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Palette {
+    Basic(BasicPalette),
+    Multi(MultiPalette),
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum BasicPalette {
+    Hot,
+    Mem,
+    Io,
+    Red,
+    Green,
+    Blue,
+    Aqua,
+    Yellow,
+    Purple,
+    Orange,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum MultiPalette {
+    Java,
+    Js,
+    Perl,
+    Wakeup,
+}
+
+impl Default for Palette {
+    fn default() -> Self {
+        Palette::Basic(BasicPalette::Hot)
+    }
+}
+
+impl FromStr for Palette {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "hot" => Ok(Palette::Basic(BasicPalette::Hot)),
+            "mem" => Ok(Palette::Basic(BasicPalette::Mem)),
+            "io" => Ok(Palette::Basic(BasicPalette::Io)),
+            "wakeup" => Ok(Palette::Multi(MultiPalette::Wakeup)),
+            "java" => Ok(Palette::Multi(MultiPalette::Java)),
+            "js" => Ok(Palette::Multi(MultiPalette::Js)),
+            "perl" => Ok(Palette::Multi(MultiPalette::Perl)),
+            "red" => Ok(Palette::Basic(BasicPalette::Red)),
+            "green" => Ok(Palette::Basic(BasicPalette::Green)),
+            "blue" => Ok(Palette::Basic(BasicPalette::Blue)),
+            "aqua" => Ok(Palette::Basic(BasicPalette::Aqua)),
+            "yellow" => Ok(Palette::Basic(BasicPalette::Yellow)),
+            "purple" => Ok(Palette::Basic(BasicPalette::Purple)),
+            "orange" => Ok(Palette::Basic(BasicPalette::Orange)),
+            unknown => Err(format!("unknown color palette: {}", unknown)),
+        }
+    }
+}
+
+struct NamehashVariables {
+    vector: f32,
+    weight: f32,
+    max: f32,
+    modulo: u8,
+}
+
+impl NamehashVariables {
+    fn init() -> Self {
+        NamehashVariables {
+            vector: 0.0,
+            weight: 1.0,
+            max: 1.0,
+            modulo: 10,
+        }
+    }
+
+    fn update(&mut self, character: u8) {
+        let i = f32::from(character % self.modulo);
+        self.vector += (i / f32::from(self.modulo - 1)) * self.weight;
+        self.modulo += 1;
+        self.max += self.weight;
+        self.weight *= 0.70;
+    }
+
+    fn result(&self) -> f32 {
+        (1.0 - self.vector / self.max)
+    }
+}
+
+/// Generate a vector hash for the name string, weighting early over
+/// later characters. We want to pick the same colors for function
+/// names across different flame graphs.
+fn namehash<I: Iterator<Item = u8>>(mut name: I) -> f32 {
```

Comment:
A couple of unit tests for this function would probably be a good idea. Let's specifically double-check what the same perl code gives to make sure we're not messing anything up!

Reply:
[Done](https://github.com/jonhoo/inferno/pull/32/commits/3f11a0397b84fb20770cc66c7659bbac2cd919f7).
Which helped fix [this bug](https://github.com/jonhoo/inferno/pull/32/commits/eba2d52a6cae571ad94f6c9f69eb95ee1fd2e901).

This is not exactly the expected values in the Perl implementation, because perl floats are double-precision values.

Reply:
I'd be okay with moving our implementation to also use `f64` if you think that change is worthwhile for consistency with `flamegraph.pl`?

Reply:
I don't think this is needed. The double precision is basically useless in this use case0